ext/nio4r/org/nio4r/Selector.java

```diff
@@ -136,7 +137,14 @@ public IRubyObject register(ThreadContext context, IRubyObject io, IRubyObject i
     @JRubyMethod
     public IRubyObject deregister(ThreadContext context, IRubyObject io) {
         Ruby runtime = context.getRuntime();
-        Channel rawChannel = RubyIO.convertToIO(context, io).getChannel();
+        Channel rawChannel;
+
+        try {
+            rawChannel = RubyIO.convertToIO(context, io).getChannel();
+        } catch(IOError e) {
```

Comment:
Looking at this implementation, I'm also now worried that this approach might "leak" [SelectionKey](http://tutorials.jenkov.com/java-nio/selectors.html#selectionkey) objects which need to get "cancelled".

@headius @enebo is there a better way to get the channel here in cases where we still need need it even after the object is closed? (i.e. to get the associated SelectionKey and properly cancel it)

Reply:
The problem is that we don't seem to have the keys anymore after the socket has been closed. So we can't do it here afaik.

Perhaps we should clear out closed sockets before select, similar to this: https://github.com/socketry/nio4r/blob/88328b085f1dbfa8de67b2fabd607f707357ae0b/ext/nio4r/org/nio4r/Selector.java#L233

This is more difficult than it seemed and I'm not familiar with Java so if that needs to be done I think I'll back off on this PR.

Reply:
The problem is `cancelKeys` is going through a list of `SelectionKey`s which gets populated, among other places, inside of the `deregister` method.